src/lib/uri/uri.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "uri.h"
+#include "uri_parser.h"
+#include "trivia/util.h"
+
+#define XSTRNDUP(s, n)  (s != NULL ? xstrndup(s, n) : NULL)
+
+void
+uri_destroy(struct uri *uri)
+{
+	free(uri->scheme);
+	free(uri->login);
+	free(uri->password);
+	free(uri->host);
+	free(uri->service);
+	free(uri->path);
+	free(uri->query);
+	free(uri->fragment);
+	TRASH(uri);
+}
+
+int
+uri_create(struct uri *uri, const char *str)
+{
+	memset(uri, 0, sizeof(struct uri));
+	if (str == NULL)
+		return 0;
+	struct uri_raw uri_raw;
```

Comment:
But `uri_raw` isn't a uri "definition", in contrast to `key_def`, `space_def`, etc, - it's more like a temporary buffer storing a parsed uri. IMO `uri_raw` is a good name. I suggest leaving it as is, because it's only used in uri internals - uri users will never deal with it.